File: src/main/java/com/abysalto/middleware/dto/PagedResponse.java
package com.abysalto.middleware.dto;

import io.swagger.v3.oas.annotations.media.Schema;

import java.util.List;

/**
 * Generic pagination envelope returned by the list/filter/search endpoints.
 *
 * @param items      the items on the current page
 * @param page       zero-based page index
 * @param size       requested page size
 * @param totalItems total number of matching items across all pages
 * @param totalPages total number of pages for the given size
 */
@Schema(description = "Paginated response envelope")
public record PagedResponse<T>(
        List<T> items,
        int page,
        int size,
        long totalItems,
        int totalPages
) {

    public static <T> PagedResponse<T> of(List<T> items, int page, int size, long totalItems) {
        int totalPages = size <= 0 ? 0 : (int) Math.ceil((double) totalItems / size);
        return new PagedResponse<>(items, page, size, totalItems, totalPages);
    }
}

File: dotnet/src/Middleware.Core/Dtos/PagedResponse.cs
namespace Middleware.Core.Dtos;

/// <summary>
/// Generic pagination envelope returned by the list/filter/search endpoints.
/// </summary>
/// <param name="Items">the items on the current page</param>
/// <param name="Page">zero-based page index</param>
/// <param name="Size">requested page size</param>
/// <param name="TotalItems">total number of matching items across all pages</param>
/// <param name="TotalPages">total number of pages for the given size</param>
public sealed record PagedResponse<T>(
    IReadOnlyList<T> Items,
    int Page,
    int Size,
    long TotalItems,
    int TotalPages)
{
    public static PagedResponse<T> Of(IReadOnlyList<T> items, int page, int size, long totalItems)
    {
        int totalPages = size <= 0 ? 0 : (int)Math.Ceiling((double)totalItems / size);
        return new PagedResponse<T>(items, page, size, totalItems, totalPages);
    }
}
